### ISSUE TITLE
Further Validate Cron Expression in Parser

### ISSUE THREAD

**Issue by amarandon** [jshmrtn/crontab#58]
`CronExpression.Parser.parse!/1` does some checks on a crontab's format, but still accepts some input that is then not usable for example by `Crontab.Scheduler.get_next_run_date/2`.

```elixir
iex(2)> alias Crontab.{CronExpression.Parser, Scheduler}                                                   
[Crontab.CronExpression.Parser, Crontab.Scheduler]
iex(3)> Parser.parse!("* */8 * * *")
        |> Scheduler.get_next_run_date(NaiveDateTime.utc_now())               
{:ok, ~N[2019-06-03 16:00:00]}
iex(4)> Parser.parse!("* 4/8 * * *")
        |> Scheduler.get_next_run_date(NaiveDateTime.utc_now())
{:error, "No compliant date was found for your interval."}
iex(5)> Parser.parse!("-1 */8 * * *")
        |> Scheduler.get_next_run_date(NaiveDateTime.utc_now())
{:error, "No compliant date was found for your interval."}
iex(6)> Parser.parse!("59 */8 * * *")
        |> Scheduler.get_next_run_date(NaiveDateTime.utc_now())
{:ok, ~N[2019-06-03 16:59:00]}
iex(7)> Parser.parse!("60 */8 * * *")
        |> Scheduler.get_next_run_date(NaiveDateTime.utc_now())
{:error, "No compliant date was found for your interval."}
```

What would be a reasonable strategy to check that a crontab coming from user input is going to be usable? I looked into the ecto data type but it seems to just use `parse`. Would it make sense to strengthen validation performed by parse? Or would the right approach be for validation code to try a call to `Scheduler.get_next_run_date/2` with the current time? Or another approach that I can't think of?

**Comment by maennchen** [comment]
If you want to validate user input, I'd recommend checking via `get_next_run_date`.

It should however not happen inside of `parse`. 

For example, a valid cron expression could specify the year `2010`. You'll never find a future run date though.

We should, however, check that the cron expression itself is valid. You can for example not have a second 60. Therefore this should yield an error.

**Comment by amarandon** [comment]
So if I understand correctly, `parse` should be improved so that it rejects things like negative numbers, minutes greater than 59, hours greater than 23 and so on? Is this something you'd be interested in adding to the library?

The year example you came up with is interesting. Let's say today a user stores a crontab with year 2019 in a database record. If I rely on `get_next_run_date` for validation based on the current date and if next year they edit the entry and try to save it back, it will give them an error even though they haven't updated the crontab field. I'm not sure if this is what we want.

Anyway in my use case I don't use the extended format so I have don't have years. I think constraining the first 5 fields to valid value ranges would allow to cover a good share of validation use cases without depending on when the validation is performed. FWIW I checked the behavior of `crontab -e` on my Linux system and it rejects any crontab with out-of-range values at edit time.

If adding value range checks to `parse` is something you'd like to see integrated in the library, I can give a go at a PR.

**Comment by maennchen** [comment]
> Is this something you'd be interested in adding to the library?

Yes, definitely.

> The year example you came up with is interesting. 

Yes, also a cron expression can be valid without being able to run in the future.

> Anyway in my use case I don't use the extended format so I have don't have years.

If you refer to `extended` of this library, then extended means including seconds and not years. If you do not further check a string that a user provides this library will always provide the option to restrict years.

> If adding value range checks to parse is something you'd like to see integrated in the library, I can give a go at a PR.

That would be great ❤️

**Comment by maennchen** [comment]
@amarandon I did some improvements to the types (dialyzer). This should help you when doing a PR:

https://github.com/jshmrtn/crontab/tree/improve_types

**Comment by amarandon** [comment]
@maennchen OK so I guess I'll rebase my current work on this then.

Actually I was wondering about how to check my code against typespecs as I've never used them before. So I should use dialyzer?

**Comment by maennchen** [comment]
@amarandon You should be able to run `mix dialyzer` and get no errors. It'd be good if you can check it and write `@spec` for all new functions as well.

**Comment by maennchen** [comment]
Closed in favor of PR #59
